src/Microsoft.CSharp/src/Microsoft/CSharp/RuntimeBinder/Syntax/NameManager.cs

```diff
@@ -2,35 +2,168 @@
+            new Name("LeftShift"),
+            new Name("RightShift"),
+            new Name("AndAlso"),
+            new Name("OrElse"),
+            new Name("Equal"),
+            new Name("NotEqual"),
+            new Name("GreaterThanOrEqual"),
+            new Name("GreaterThan"),
+            new Name("LessThan"),
+            new Name("LessThanOrEqual"),
+            new Name("ArrayIndex"),
+            new Name("Assign"),
+            new Name("Condition"),
+            new Name("Field"),
+            new Name("Call"),
+            new Name("New"),
+            new Name("Quote"),
+            new Name("ArrayLength"),
+            new Name("UnaryPlus"),
+            new Name("Negate"),
+            new Name("NegateChecked"),
+            new Name("Not"),
+            new Name("NewArrayInit"),
+            new Name("Property"),
+            new Name("AddEventHandler"),
+            new Name("RemoveEventHandler"),
+            new Name("InvocationList"),
+            new Name("GetOrCreateEventRegistrationTokenTable"),
+            /* Above here corresponds with PredefinedName enum */
+            new Name("true"),
+            new Name("false"),
+            new Name("null"),
+            new Name("base"),
+            new Name("this"),
+            new Name("explicit"),
+            new Name("implicit"),
+            new Name("__arglist"),
+            new Name("__makeref"),
+            new Name("__reftype"),
+            new Name("__refvalue"),
+            new Name("as"),
+            new Name("checked"),
+            new Name("is"),
+            new Name("typeof"),
+            new Name("unchecked"),
+            new Name("void"),
+            new Name("")
+        };
 
-        internal NameManager()
-            : this(new NameTable())
-        {
-        }
+        private static readonly NameTable s_knownNames = GetKnownNames();
 
-        private NameManager(NameTable nameTable)
+        private static NameTable GetKnownNames()
         {
-            _names = nameTable;
-            InitKnownNames();
+            NameTable table = new NameTable();
+
+            foreach (Name name in s_predefinedNames)
+            {
+                table.Add(name);
+            }
+
+            return table;
         }
 
+        private readonly NameTable _names = new NameTable();
+
         internal Name Add(string key)
         {
             if (key == null)
             {
                 throw Error.InternalCompilerError();
             }
-            Name name = s_knownNames.Lookup(key);
-            if (name == null)
-            {
-                name = _names.Add(key);
-            }
-            return name;
+
+            return s_knownNames.Lookup(key) ?? _names.Add(key);
+        }
+
+        internal Name Add(string key, int length)
+        {
+            Debug.Assert(key != null);
+            Debug.Assert(key.Length >= length);
+            return s_knownNames.Lookup(key, length) ?? _names.Add(key, length);
```

Comment:
it seems  `s_knownNames.Lookup(key, length)` might never return anything since known names are not for anything generic. 
If we do not need to support `Lookup(key, length)` for knownNames, it could be just a string switch, which would be more efficient and not require any initialization.